humblegen/src/rust.rs

```diff
@@ -150,9 +162,389 @@ fn render_atom(atom: &ast::AtomType) -> TokenStream {
+            (
+                quote! { post_body },
+                quote! { },
+            )
+        }));
+        let post_body_def = r.post_body_type.as_ref().map(|pbt| quote!{
+            let post_body: #pbt =
+            deser_post_data(req.body_mut()).await?;
+        });
+        // query
+        arg_list_and_defs.extend(r.query_type.iter().map(|_| {
+            (
+                quote!{ query },
+                quote!{},
+            )
+        }));
+        let query_deser_fn = &r.query_deser_fn;
+        let query_def = r.query_type.as_ref().map(|qt| quote!{
+            let query: Option<#qt> = match req.uri().query() {
+                None => None,
+                Some(q) => Some(#query_deser_fn(q)?),
+            };
+        });
+        // route params
+        arg_list_and_defs.extend(r.components.iter().filter_map(|c| match c {
+            ServiceRouteComponent::Literal { .. } => None,
+            ServiceRouteComponent::Param {
+                spec_arg_name,
+                rust_var_ident,
+                rust_var_type,
+                rust_deser_fn,
+                ..
+            } => Some((
+                quote! { #rust_var_ident },
+                quote! {
+                    let #rust_var_ident: #rust_var_type =
+                        #rust_deser_fn (&captures[ #spec_arg_name ]) ;},
+            )),
+        }));
+        let (arg_list, arg_defs): (Vec<TokenStream>, Vec<TokenStream>) =
+            arg_list_and_defs.into_iter().unzip();
+        let (arg_list, arg_defs) = (arg_list.into_iter(), arg_defs.into_iter());
+
+        quote! {
+            {
+                let handler = Arc::clone(&handler);
+                Route{
+                    method: #hyper_method,
+                    regex: ::humblegen_rt::regex::Regex::new(#regex_str).unwrap(),
+                    dispatcher: Box::new(
+                        move |mut req: ::humblegen_rt::hyper::Request<::humblegen_rt::hyper::Body>,
+                        captures| {
+                            let handler = Arc::clone(&handler);
+                            #(#arg_defs);*
+                            Box::pin(async move {
+                                #query_def
+                                #post_body_def
+                                Ok(handler_response_to_hyper_response(handler.#traitfn_ident( #(#arg_list),* ).await))
+                            })
+                        }
+                    ),
+                }
+            }
+        }
+    });
+
+    let routes_factory_name = &service.routes_factory_name;
+    quote! {
+        #trait_def
+
+        #[allow(unused_variables)]
+        #[allow(unused_mut)]
+        #[allow(non_snake_case)]
+        #[allow(clippy::trivial_regex)]
+        #[allow(clippy::single_char_pattern)]
+        fn #routes_factory_name(handler: Arc<dyn #trait_name + Send + Sync>) -> Vec<Route> {
+            vec![#(#routes),*]
+        }
+
+    }
+}
+
+struct ServiceRoute {
```

Comment:
Do all the structs we used here not allow derivation of `Debug`? I would want that on every `struct` where it is free.

Reply:
not quite free, but did it now anyways